### PR TITLE
Sap lib + tfstate deployment

### PR DESCRIPTION
**Pull request by YunziZhangAzure**
### Problem
Currently, tfstates are stored in local, which makes them vulnerable. 

### Solution
Seperate each module.
This pr focus on sap library: provision storage accounts and manage their tfstate.
/bootstrap/sap_library will be used to provision storage accounts for storing sapbits and tfstates
/run/sap_library will be used to config azurem as the backend of terraform and help to migrate the tfstates of storage accounts from local to remote storage account.

### Test
1. Logon to deployer vm, (checkout to this git branch) cd ~/sap-hana/deploy/terraform/bootstrap/sap_library
2. Edit sap_library.json.
3. Execute: terraform init
4. Execute: terraform apply -auto-approve -var-file=sap_library.json
5. After execution, there will be a shell script "tfstate_scripts.sh" generated.
6. Under "default" directory, execute: sh tfstate_scripts.sh 
7. Check if corresponding container has the saplibrary.terraform.tfstate file.

### CODE REVIEW

File: deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+// Generates random text for boot diagnostics storage account name
+resource "random_id" "random-id" {
+  keepers = {
+    # Generate a new id only when a new resource group is defined
+    resource_group = local.rg_exists ? data.azurerm_resource_group.library[0].name : azurerm_resource_group.library[0].name
+  }
+  byte_length = 4
+}
```
Comment (mkdeegan) [inline deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf:100]: Why? Boot Diagnostics Storage account should be created with the Deploy VNET and the SAP Workload VNET.

File: deploy/terraform/terraform-units/modules/sap_library/outputs.tf [hunk 2]
```diff
@@ -0,0 +1,9 @@
+output rgName { 
+    value = azurerm_resource_group.library[0].name 
+}
+output tfstateBlobEndpoint {
+    value = azurerm_storage_account.storage-tfstate[0].primary_blob_endpoint 
+}
+output sapbitsBlobEndpoint { 
+    value = azurerm_storage_account.storage-sapbits[0].primary_blob_endpoint 
+}
```
Comment (mkdeegan) [inline deploy/terraform/terraform-units/modules/sap_library/outputs.tf:9]: This should not be an output here.

File: deploy/terraform/terraform-units/modules/sap_library/infrastructure.tf [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+// Imports data of existing mgmt subnet
+data "azurerm_subnet" "subnet-mgmt" {
+  count                = local.sub_mgmt_exists ? 1 : 0
+  name                 = split("/", local.sub_mgmt_arm_id)[10]
+  resource_group_name  = split("/", local.sub_mgmt_arm_id)[4]
+  virtual_network_name = split("/", local.sub_mgmt_arm_id)[8]
+}
+
```
Comment (mkdeegan) [inline deploy/terraform/terraform-units/modules/sap_library/infrastructure.tf:50]: No VNETS or Subnet Creation allowed as part of this Unit of work

File: deploy/terraform/initialize/sap_library_input.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    },
+    "storage_account_tfstate": {
+        "is_existing": "",
+        "arm_id": "",
+        "account_tier": "",
+        "account_replication_type": "",
+        "account_kind": "",
+        "container_access_type": ""
+    }
+}
```
Comment (mkdeegan) [inline deploy/terraform/initialize/sap_library_input.json:49]: We should not convert to JSON since it already was operational in the desired state of TFVARS. We will only have to cover back to support TF Cloud and Enterprise.

File: deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  Description:
+  Set up storage accounts for sap library 
+*/
+
+// Imports existing storage account to use for tfstate
+data "azurerm_storage_account" "storage-tfstate" {
+  count               = local.sa_tfstate_exists ? 1 : 0
+  name                = split("/", local.sa_tfstate_arm_id.arm_id)[8]
+  resource_group_name = split("/", local.sa_tfstate_arm_id.arm_id)[4]
+}
```
Comment (mkdeegan) [inline deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf:11]: These data constructs make the important code harder to read. What if we have a data.tf file and collect them all there as they are data manipulation constructs and not resource creation blocks.

File: deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf (hunk 1 above)
Comment (YunziZhangAzure) [reply deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf:100]: Thanks for reviewing my pr! This pr is a draft though. I will keep developing it, and make it ready for another review ASAP.

I will remove Boot diagnostics from this module.

File: deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf (hunk 5 above)
Comment (YunziZhangAzure) [reply deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf:11]: The principal I'm using to name these resources is by resource type, by functionality.
The reason why I put it here is because it is about storage_accounts.tf. If there's something we want to change in the future, we can search for the name "storage_accounts" directly. However, by looking at the 'data.tf', we don't know what it means.

File: deploy/terraform/terraform-units/modules/sap_library/infrastructure.tf (hunk 3 above)
Comment (YunziZhangAzure) [reply deploy/terraform/terraform-units/modules/sap_library/infrastructure.tf:50]: Got it! Will remove this part.

File: deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf (hunk 5 above)
Comment (mkdeegan) [reply deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf:11]: Hi Yunzi...
Just in reading this, the data construct is only used when the JSON exists and the name and resource group are provided in an ARM ID. This type of code has nothing to do with the creation of the resource and is more like a pseudo construct for a customer provided override. Moving them into a separate file would logically separate the Resource Import functionality from the Resource create. It is only a thought for easier readability... functionally it does not matter. Do what ever makes it more concise and easier to understand.... It is just a log to noise for the customer to look in a code review when its only function is to parse a JSON element.

File: deploy/terraform/terraform-units/modules/sap_library/infrastructure.tf [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  Set up infrastructure for sap library 
+*/
+
+resource "azurerm_resource_group" "library" {
+  count    = local.rg_exists ? 0 : 1
+  name     = local.rg_name
+  location = local.region
+}
+
+# Imports data of existing resource group
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_library/infrastructure.tf:12]:
```suggestion
// Imports data of existing resource group
```

File: deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+resource "azurerm_storage_share" "fileshare-sapbits" {
+  count                = local.sa_sapbits_exists ? 0 : (local.sa_sapbits_file_share_name == "" ? 0 : 1)
+  name                 = local.sa_sapbits_file_share_name
+  storage_account_name = azurerm_storage_account.storage-sapbits[0].name
+}
+
+// Generates random text for storage account name
+resource "random_id" "random-id" {
+  keepers = {
+    # Generate a new id only when a new resource group is defined
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf:94]:
```suggestion
    // Generate a new id only when a new resource group is defined
```

File: deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+// Creates file share inside the storage account for SAP bits
+resource "azurerm_storage_share" "fileshare-sapbits" {
+  count                = local.sa_sapbits_exists ? 0 : (local.sa_sapbits_file_share_name == "" ? 0 : 1)
+  name                 = local.sa_sapbits_file_share_name
+  storage_account_name = azurerm_storage_account.storage-sapbits[0].name
+}
+
+// Generates random text for storage account name
+resource "random_id" "random-id" {
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf:92]: Change `random-id` to something like post-fix would make it more readable :)

File: deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+// Creates the storage container inside the storage account for deployer
+resource "azurerm_storage_container" "storagecontainer-deployer" {
+  count                 = local.sa_tfstate_exists ? 0 : 1
+  name                  = local.sa_deployer_container_name
+  storage_account_name  = local.sa_tfstate_exists ? data.azurerm_storage_account.storage-tfstate[0].name : azurerm_storage_account.storage-tfstate[0].name
+  container_access_type = local.sa_tfstate_container_access_type
+}
+
+// Creates the storage container inside the storage account for saplibrary
+resource "azurerm_storage_container" "storagecontainer-saplibrary" {
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf:50]: Consider adding `delete_retention_policy` for blob soft deletion.

File: deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+// Imports existing storage account to use for SAP bits
+data "azurerm_storage_account" "storage-sapbits" {
+  count               = local.sa_sapbits_exists ? 1 : 0
+  name                = split("/", local.sa_sapbits_arm_id.arm_id)[8]
+  resource_group_name = split("/", local.sa_sapbits_arm_id.arm_id)[4]
+}
+
+// Creates storage account for storing SAP Bits
+resource "azurerm_storage_account" "storage-sapbits" {
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_library/storage_accounts.tf:65]: Same as above.

File: deploy/terraform/terraform-units/modules/sap_library/infrastructure.tf [hunk 11]
```diff
@@ -0,0 +1,16 @@
+/*
+  Description:
+  Set up infrastructure for sap library 
+*/
+
+resource "azurerm_resource_group" "library" {
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_library/infrastructure.tf:6]: Consider adding `azurerm_management_lock` to protect the resources with lock.

File: util/terraform_v2.sh [hunk 12]
```diff
@@ -28,7 +28,6 @@ readonly input_file_term='<JSON template name>'
 
 readonly target_code="${target_path}/terraform/run/sap_landscape"
 
-
```
Comment (rukawata) [inline util/terraform_v2.sh:31]: Why this change in this PR?

File: deploy/terraform/bootstrap/sap_library/providers.tf [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    >    greater than version number
+    >=   greater than or equal to version number
+    <    less than version number
+    <=   less than or equal to version number
+    ~>   pessimistic constraint operator, constraining both the oldest and newest version allowed.
+           For example, ~> 0.9   is equivalent to >= 0.9,   < 1.0 
+                        ~> 0.8.4 is equivalent to >= 0.8.4, < 0.9
+*/
+
+provider "azurerm" {
```
Comment (rukawata) [inline deploy/terraform/bootstrap/sap_library/providers.tf:16]: `use_msi = true` is missing because we use MSI to run `terraform apply`.
Other than that, something else is not setup properly.
I manually add `use_msi = true` to the `providers.tf`, but I still can't run terraform apply after logon with MSI.
```
azureadm@deployer00-vm-f69ad65b:~/sap-hana/deploy/terraform/bootstrap/sap_library$ az login --identity
[
  {
    "environmentName": "AzureCloud",
    "homeTenantId": "72f988bf-86f1-41af-91ab-2d7cd011db47",
    "id": "c4106f40-4f28-442e-b67f-a24d892bf7ad",
    "isDefault": true,
    "managedByTenants": [
      {
        "tenantId": "2f4a9838-26b7-47ee-be60-ccc1fdec5953"
      }
    ],
    "name": "AzLinux SAP Landscapes",
    "state": "Enabled",
    "tenantId": "72f988bf-86f1-41af-91ab-2d7cd011db47",
    "user": {
      "assignedIdentityInfo": "MSI",
      "name": "systemAssignedIdentity",
      "type": "servicePrincipal"
    }
  }
]
azureadm@deployer00-vm-f69ad65b:~/sap-hana/deploy/terraform/bootstrap/sap_library$ terraform apply -auto-approve -var-file=sap_library.json

Error: Unable to list provider registration status, it is possible that this is due to invalid credentials or the service principal does not have permission to use the Resource Manager API, Azure error: resources.ProvidersClient#List: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="SubscriptionNotFound" Message="The subscription 'providers' could not be found."

  on providers.tf line 16, in provider "azurerm":
  16: provider "azurerm" {
```

File: deploy/terraform/run/sap_library/providers.tf [hunk 14]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (rukawata) [inline deploy/terraform/run/sap_library/providers.tf:1]: I assume we will have a separate file for different backend? local vs. azurerm.

File: util/terraform_v2.sh (hunk 12 above)
Comment (YunziZhangAzure) [reply util/terraform_v2.sh:31]: I suspect it is because the this file is changed accidentally before. so I changed it back.

File: deploy/terraform/run/sap_landscape/module.tf [hunk 15]
```diff
@@ -3,7 +3,7 @@
   Setup common infrastructure
 */
 module "common_infrastructure" {
-  source              = "../modules/common_infrastructure"
+  source              = "./../modules/common_infrastructure"
```
[the remark this answers is not in the document]
Comment (YunziZhangAzure) [reply deploy/terraform/run/sap_landscape/module.tf:6]: Just for being more clear. The functionality is same I think. If you want me to keep the original version, I can change it back.

File: deploy/terraform/bootstrap/sap_library/scripts.tf [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+  Generate a script to migrate tfstate of storage accounts from local to remote
+*/
+resource "local_file" "output-script" {
+    content = templatefile("${path.module}/tfstate_scripts.tmpl", {
+    saplibrary-resource-group       =  module.sap_library.rgName
+    tfstate-storage-account-name    =  module.sap_library.tfstate-storage-account-name
+    }
+  )
+  filename             = "${path.root}/tfstate_scripts.sh"
+  file_permission      = "0660"
```
Comment (rukawata) [inline deploy/terraform/bootstrap/sap_library/scripts.tf:12]: Can we generate the file under workspace path` ${terraform.workspace}`? which would be `default` if not set.
